### PR TITLE
Remove special `any` assignability for numeric index signatures

### LINKED ISSUE #18757: Functions are compatible with ArrayLike<any>
**TypeScript Version:**  2.5.2

**Code**

```ts
function func() {}
const array: ArrayLike<any> = func;
```

See [this TypeScript Playground example](http://bit.ly/2hvrZq2).

**Expected behavior:**

I'd have expected an error to be effected.

**Actual behavior:**

No error is effected if the type `ArrayLike<any>` is specified. It seems the absence of an index signature is not a consideration in this situation.

If the `any` is changed to, say, `string` an error is effected:

    Index signature is missing in type '() => void'.

### PR DESCRIPTION
**Pull request by DanielRosenwasser**
Experiment to fix #18757

### CODE REVIEW

**Comment by DanielRosenwasser** [comment]
@typescript-bot pack this
@typescript-bot test this
@typescript-bot user test this
@typescript-bot run dt
@typescript-bot perf test this

**Comment by DanielRosenwasser** [comment]
A couple of breaks:

* One of the tests added as part of https://github.com/microsoft/TypeScript/pull/29179
* Office UI Fabric
  * https://github.com/typescript-bot/TypeScript/pull/109/files#diff-02fa25d51ee84a35829f43b04a84da8bce4668eed9b38642fc37504afa10004bR789
  * https://github.com/typescript-bot/TypeScript/pull/109/files#diff-02fa25d51ee84a35829f43b04a84da8bce4668eed9b38642fc37504afa10004bR1025

File: tests/baselines/reference/subclassThisTypeAssignable.errors.txt [hunk 1]
```diff
@@ -1,9 +1,23 @@
...
+tests/cases/compiler/tile1.ts(24,7): error TS2322: Type 'C' is not assignable to type 'ClassComponent<any>'.
+  Index signature is missing in type 'C'.
 
 
-==== tests/cases/compiler/tile1.ts (1 errors) ====
+==== tests/cases/compiler/tile1.ts (5 errors) ====
     interface Lifecycle<Attrs, State> {
     	oninit?(vnode: Vnode<Attrs, State>): number;
+    	                            ~~~~~
+!!! error TS2344: Type 'State' does not satisfy the constraint 'Lifecycle<Attrs, State>'.
```
Comment (DanielRosenwasser) [inline tests/baselines/reference/subclassThisTypeAssignable.errors.txt:20]: Honestly, this is a bug.

File: tests/baselines/reference/subclassThisTypeAssignable.errors.txt [hunk 2]
```diff
@@ -25,10 +41,23 @@ tests/cases/compiler/tile1.ts(6,81): error TS2744: Type parameter defaults can o
     interface MyAttrs { id: number }
     class C implements ClassComponent<MyAttrs> {
     	view(v: Vnode<MyAttrs>) { return 0; }
+    	~~~~
+!!! error TS2416: Property 'view' in type 'C' is not assignable to the same property in base type 'ClassComponent<MyAttrs>'.
```
Comment (DanielRosenwasser) [inline tests/baselines/reference/subclassThisTypeAssignable.errors.txt:45]: This basically comes down to the following break:

```ts
interface Indexable {
    [x: number]: any;
}

class C implements Indexable {
}
```

In earlier versions, implementing `C` without re-declaring the index signature was valid. I would argue that this behavior is still surprising or incorrect for string index signatures.

**Comment by DanielRosenwasser** [comment]
I still need to add tests for the original use-case.

**Comment by DanielRosenwasser** [comment]
Updates:

* The original use-case is now in our test suite. A function will not be assignable to an `ArrayLike<any>`.
* I added a modified version of `subclassThisTypeAssignable` to the test suite that has no errors at all. It avoids weird circularity errors, appropriately adds constraints where needed, and adds the index signature that this PR now requires.

**Comment by DanielRosenwasser** [comment]
@typescript-bot pack this
@typescript-bot test this
@typescript-bot user test this
@typescript-bot run dt
@typescript-bot perf test this faster

**Comment by DanielRosenwasser** [comment]
I don't think we're t correctly running the community test suite anymore (https://github.com/microsoft/TypeScript/issues/42875), so I can't confidently comment on if anything new broke. But I think this is now a "serious PR".

**Comment by DanielRosenwasser** [comment]
@typescript test this

**Comment by RyanCavanaugh** [comment]
Assuming there's nothing too concerning in the user tests, I'm okay with putting this in the nightly and watching for fallout. I stand by the current behavior being a bug 🙂

**Review by RyanCavanaugh — APPROVED** [review]

**Pull request merged**